[Bug]: Go Validates Runner tests failing under Flink 2.0

### What happened?

See https://github.com/apache/beam/pull/37313#issuecomment-3899021609

```
com.esotericsoftware.kryo.KryoException: java.lang.UnsupportedOperationException
Serialization trace:
fieldIndices (org.apache.beam.sdk.schemas.Schema)
schema (org.apache.beam.sdk.values.RowWithStorage)
value (org.apache.beam.sdk.values.KV)
value (org.apache.beam.sdk.values.WindowedValues$ValueInGlobalWindow)
        at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:146)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:130)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:796)
        at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:124)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:130)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:796)
        at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:124)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:130)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:796)
        at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:124)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:130)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:877)
        at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.deserialize(KryoSerializer.java:384)
        at org.apache.flink.streaming.runtime.streamrecord.StreamElementSerializer.deserialize(StreamElementSerializer.java:206)
        at org.apache.flink.streaming.runtime.streamrecord.StreamElementSerializer.deserialize(StreamElementSerializer.java:44)
        at org.apache.flink.runtime.plugable.NonReusingDeserializationDelegate.read(NonReusingDeserializationDelegate.java:53)
        at org.apache.flink.runtime.io.network.api.serialization.NonSpanningWrapper.readInto(NonSpanningWrapper.java:337)
        at org.apache.flink.runtime.io.network.api.serialization.SpillingAdaptiveSpanningRecordDeserializer.readNonSpanningRecord(SpillingAdaptiveSpanningRecordDeserializer.java:130)
        at org.apache.flink.runtime.io.network.api.serialization.SpillingAdaptiveSpanningRecordDeserializer.readNextRecord(SpillingAdaptiveSpanningRecordDeserializer.java:105)
        at org.apache.flink.runtime.io.network.api.serialization.SpillingAdaptiveSpanningRecordDeserializer.getNextRecord(SpillingAdaptiveSpanningRecordDeserializer.java:95)
        at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.emitNext(AbstractStreamTaskNetworkInput.java:152)
        at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:65)
        at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:638)
        at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:231)
        at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:980)
        at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:917)
        at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:963)
        at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:942)
        at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:756)
        at org.apache.flink.runtime.taskmanager.Task.run(Task.java:568)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.UnsupportedOperationException
        at org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap.put(ImmutableMap.java:781)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:236)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:42)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:796)
        at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:124)
        ... 30 more

--- FAIL: TestTimers_EventTime_Bounded (10.32s)
FAIL
FAIL    github.com/apache/beam/sdks/v2/go/test/integration/primitives   11.168s
FAIL
```

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Infrastructure
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

--- a/.github/trigger_files/beam_PostCommit_Go_VR_Flink.json
+++ b/.github/trigger_files/beam_PostCommit_Go_VR_Flink.json
@@ -2,4 +2,5 @@
   "comment": "Modify this file in a trivial way to cause this test suite to run",
   "modification": 2,
   "https://github.com/apache/beam/pull/32440": "testing datastream optimizations",
+  "pr": "37640"
 }

--- a/runners/flink/2.0/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java
+++ b/runners/flink/2.0/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java
@@ -36,6 +36,7 @@ import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Stream
 import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.net.HostAndPort;
 import org.apache.flink.api.common.ExecutionConfig;
 import org.apache.flink.api.common.RuntimeExecutionMode;
+import org.apache.flink.api.common.serialization.SerializerConfigImpl;
 import org.apache.flink.configuration.CheckpointingOptions;
 import org.apache.flink.configuration.Configuration;
 import org.apache.flink.configuration.CoreOptions;
@@ -270,6 +271,7 @@ public class FlinkExecutionEnvironments {
       flinkStreamEnv.getConfig().setAutoWatermarkInterval(options.getAutoWatermarkInterval());
     }
     configureWebUIOptions(flinkStreamEnv.getConfig(), options.as(PipelineOptions.class));
+    configureCustomKryoSerializers(flinkStreamEnv.getConfig());
 
     return flinkStreamEnv;
   }
@@ -292,6 +294,14 @@ public class FlinkExecutionEnvironments {
     } catch (Exception e) {
       LOG.warn("Unable to configure web ui options", e);
     }
+  }
+
+  private static void configureCustomKryoSerializers(ExecutionConfig config) {
+    SerializerConfigImpl serializerConfig = (SerializerConfigImpl) config.getSerializerConfig();
+    // Force Beam schema to use JavaSerializer to fix serialization involving ImmutableMap
+    serializerConfig.registerTypeWithKryoSerializer(
+        org.apache.beam.sdk.schemas.Schema.class,
+        com.esotericsoftware.kryo.serializers.JavaSerializer.class);
   }
 
   private static class GlobalJobParametersImpl extends ExecutionConfig.GlobalJobParameters {

--- a/runners/flink/flink_runner.gradle
+++ b/runners/flink/flink_runner.gradle
@@ -230,6 +230,13 @@ dependencies {
   // configuration (https://issues.apache.org/jira/browse/BEAM-11732).
   permitUnusedDeclared "org.apache.flink:flink-clients:$flink_version"
 
+  // align with Flink's kryo version (groupId changed)
+  if (flink_major.startsWith('2')) {
+    implementation "com.esotericsoftware:kryo:5.6.2"
+  } else {
+    implementation "com.esotericsoftware.kryo:kryo:2.24.0"
+  }
+
   implementation "org.apache.flink:flink-streaming-java:$flink_version"
   testImplementation "org.apache.flink:flink-statebackend-rocksdb:$flink_version"
   testImplementation "org.apache.flink:flink-streaming-java:$flink_version:tests"

--- a/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java
+++ b/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java
@@ -298,6 +298,7 @@ public class FlinkExecutionEnvironments {
     configureStateBackend(options, flinkStreamEnv);
 
     configureWebUIOptions(flinkStreamEnv.getConfig(), options.as(PipelineOptions.class));
+    configureCustomKryoSerializers(flinkStreamEnv.getConfig());
 
     return flinkStreamEnv;
   }
@@ -320,6 +321,13 @@ public class FlinkExecutionEnvironments {
     } catch (Exception e) {
       LOG.warn("Unable to configure web ui options", e);
     }
+  }
+
+  private static void configureCustomKryoSerializers(ExecutionConfig config) {
+    // Force Beam schema to use JavaSerializer to fix serialization involving ImmutableMap
+    config.registerTypeWithKryoSerializer(
+        org.apache.beam.sdk.schemas.Schema.class,
+        com.esotericsoftware.kryo.serializers.JavaSerializer.class);
   }
 
   private static class GlobalJobParametersImpl extends ExecutionConfig.GlobalJobParameters {

--- a/sdks/go/test/build.gradle
+++ b/sdks/go/test/build.gradle
@@ -79,8 +79,7 @@ task dataflowValidatesRunnerARM64() {
 task flinkValidatesRunner {
   group = "Verification"
 
-  // TODO(https://github.com/apache/beam/issues/37600) use project.ext.latestFlinkVersion after resolved
-  def flinkVersion = '1.20'
+  def flinkVersion = project.ext.latestFlinkVersion
 
   dependsOn ":sdks:go:test:goBuild"
   dependsOn ":sdks:go:container:docker"

--- a/sdks/go/test/integration/integration.go
+++ b/sdks/go/test/integration/integration.go
@@ -199,7 +199,7 @@ var flinkFilters = []string{
 	"TestTestStreamToGBK",
 	"TestTestStreamTimersEventTime",
 
-	"TestTimers_EventTime_Unbounded", // (failure when comparing on side inputs (NPE on window lookup))
+	"TestTimers_EventTime_WithNoOutputTimestamp", // Encounter error: TimestampCombiner moved element from TIMESTAMP_MAX_VALUE to earlier time (end of global window) for window GlobalWindow
 	"TestTimers_ProcessingTime.*",    // Flink doesn't support processing time timers.
 
 	// no support for BundleFinalizer

